File: 0017-letter-combinations-of-a-phone-number/c++/Solution.cpp
class Solution {
public:
    vector<string> result;
    vector<string> letterCombinations(string digits) {
        if (digits.size() == 0) return {};
        vector<string> mappings(10);
        mappings[2] = "abc";
        mappings[3] = "def";
        mappings[4] = "ghi";
        mappings[5] = "jkl";
        mappings[6] = "mno";
        mappings[7] = "pqrs";
        mappings[8] = "tuv";
        mappings[9] = "wxyz";
        vector<char> path(digits.size());
        backtrack(mappings, digits, 0, path);
        return result;
    }
    void backtrack(vector<string> mappings, string digits, int k, vector<char> path) {
        if (k == digits.size()) {
            string str(path.begin(), path.end());
            result.push_back(str);
            return;
        }
        string mapping = mappings[digits[k] - '0'];
        for (int i = 0; i < mapping.size(); ++i) {
            path[k] = mapping[i];
            backtrack(mappings, digits, k + 1, path);
        }
    }
};

File: 0017-letter-combinations-of-a-phone-number/python/Solution.py
class Solution:
    def __init__(self):
        self.result = []

    def letterCombinations(self, digits: str) -> List[str]:
        if len(digits) == 0:
            return self.result
        mappings = [None] * 10
        mappings[2] = "abc"
        mappings[3] = "def"
        mappings[4] = "ghi"
        mappings[5] = "jkl"
        mappings[6] = "mno"
        mappings[7] = "pqrs"
        mappings[8] = "tuv"
        mappings[9] = "wxyz"
        path = [None] * len(digits)
        self.backtrack(mappings, digits, 0, path)
        return self.result

    def backtrack(self, mappings, digits, k, path):
        if k == len(digits):
            self.result.append("".join(path))
            return
        mapping = mappings[int(digits[k])]
        for i in range(len(mapping)):
            path[k] = mapping[i]
            self.backtrack(mappings, digits, k+1, path)
